ev3dev.go

```diff
@@ -226,33 +287,55 @@ func Wait(d StaterDevice, mask, want, not MotorState, any bool, timeout time.Dur
 	}
 	defer f.Close()
 
-	// If any state in the mask is wanted, we just
-	// need to check that (stat^not)&mask is not zero.
-	if any {
-		want = 0
+	// See if we can exit early.
+	stat, err = d.State()
+	if err != nil {
+		return stat, false, err
+	}
+	if stateIsOK(stat, mask, want, not, any) {
+		return stat, true, nil
+	}
+
+	// The sysfs_notify hooks in the motor drivers
+	// do not monitor stalled or overloaded states,
+	// so if the user wants to wait on these, we
+	// cannot use poll.
+	var canPoll bool
+	if pollIsSupported {
+		noPollMask := mask & (Stalled | Overloaded)
+		if !any {
+			noPollMask &= want | not
+		}
+		canPoll = noPollMask == 0
+	}
+	var fds []unix.PollFd
+	if canPoll {
+		fds = []unix.PollFd{{Fd: int32(f.Fd()), Events: unix.POLLIN}}
+
+		// Read a single byte to mark f as unchanged.
+		f.ReadAt([]byte{0}, 0)
```

Comment:
On second thoughts, the raciness introduced by having two accesses to this is not nice. I'll factor out the state call somehow.